internal/jobcontainers/job.go

```diff
@@ -0,0 +1,364 @@
+package jobcontainers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+// This file provides higher level constructs for the win32 job object API.
+// Most of the core creation and management functions are already present in "golang.org/x/sys/windows"
+// (CreateJobObject, AssignProcessToJobObject, etc.) as well as most of the limit information
+// structs and associated limit flags. Whatever is not present from the job object API
+// in golang.org/x/sys/windows is located in /internal/winapi.
+//
+// https://docs.microsoft.com/en-us/windows/win32/procthread/job-objects
+
+// jobObject is a high level wrapper around a Windows job object. Holds a handle to
+// the job and a handle to an iocp to be used to receive notifications about the lifecycle
+// of the job
+type jobObject struct {
+	jobHandle  windows.Handle
+	iocpHandle windows.Handle
+	handleLock sync.RWMutex
+}
+
+type jobLimits struct {
+	affinity       uintptr
+	cpuRate        uint32
+	cpuWeight      uint32
+	jobMemoryLimit uintptr
+	maxIops        int64
+	maxBandwidth   int64
+}
+
+var errAlreadyClosed = errors.New("the handle has already been closed")
+
+func (job *jobObject) setResourceLimits(ctx context.Context, limits *jobLimits) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	// Go through and check what limits were specified and construct the appropriate
+	// structs.
+	if limits.affinity != 0 || limits.jobMemoryLimit != 0 {
+		var (
+			basicLimitFlags uint32
```

Comment:
is this actually used by anything? Looks like you set the value but don't do anything with it as far as I can tell.